h2o-core/src/main/java/hex/grid/GridSearch.java

```diff
@@ -250,8 +250,6 @@ private void attemptBuildNextModel(final ParallelModelBuilder parallelModelBuild
           reconcileMaxRuntime(grid._key, nextModelParams);
 
           parallelModelBuilder.run(Collections.singletonList(ModelBuilder.make(nextModelParams)));
-        } else {
-          parallelModelBuilder.noMoreModels();
```

Comment:
it's still needed to prevent the `grid.update()` on unlocked grid issue right?
We need a follow up ticket to fix the cancellation scenarios, and distinguish the following cases:
1. the grid is exhausted or early stopped: in this case we wait for the pending models and the parallel builder can be marked as complete only when all models are built or failed.
2. the grid is cancelled (explicit `job.stop` or timeout): in this case we should prevent new modelbuilders to be pushed, and running subjobs should be explicitely stopped as well. The parallel builder can also be marked as complete when the pending jobs are done (failed due to cancellation).

Reply:
Agreed.